bodhi/tests/server/test_metadata.py

```diff
@@ -342,48 +348,72 @@ def _test_extended_metadata(self, has_alias):
+
+            self.assertEqual(len(uinfo.updates), 1)
+            notice = uinfo.updates[0]
+
+            self.assertIsNotNone(notice)
+            self.assertEqual(notice.title, update.title)
+            self.assertEqual(notice.release, update.release.long_name)
+            self.assertEqual(notice.status, update.status.value)
+            if update.date_modified:
+                self.assertEqual(notice.updated_date, update.date_modified)
+            self.assertEqual(notice.fromstr, config.get('bodhi_email'))
+            self.assertEqual(notice.rights, config.get('updateinfo_rights'))
+            self.assertEqual(notice.description, update.notes)
+            self.assertEqual(notice.id, update.alias)
+            self.assertEqual(notice.severity, 'Moderate')
+            bug = notice.references[0]
+            self.assertEqual(bug.href, update.bugs[0].url)
+            self.assertEqual(bug.id, '12345')
+            self.assertEqual(bug.type, 'bugzilla')
+            cve = notice.references[1]
+            self.assertEqual(cve.type, 'cve')
+            self.assertEqual(cve.href, update.cves[0].url)
+            self.assertEqual(cve.id, update.cves[0].cve_id)
+
+            col = notice.collections[0]
+            self.assertEqual(col.name, update.release.long_name)
+            self.assertEqual(col.shortname, update.release.name)
+
+            pkg = col.packages[0]
+            self.assertEqual(pkg.epoch, '0')
+            self.assertEqual(pkg.name, 'TurboGears')
+            self.assertEqual(
+                pkg.src,
+                ('https://download.fedoraproject.org/pub/fedora/linux/updates/testing/17/SRPMS/T/'
+                 'TurboGears-1.0.2.2-2.fc17.src.rpm'))
+            self.assertEqual(pkg.version, '1.0.2.2')
+            self.assertFalse(pkg.reboot_suggested)
+            self.assertEqual(pkg.arch, 'src')
+            self.assertEqual(pkg.filename, 'TurboGears-1.0.2.2-2.fc17.src.rpm')
+
+    # Make sure we reach 100% coverage, even though cr.ZCK_COMPRESSION is only defined
+    # when createrepo_c supports zchunk
+    @mock.patch('bodhi.server.metadata.cr')
+    def test_zchunk_metadata(self, mock_cr):
+        mock_cr.ZCK_COMPRESSION = 99
+        assert(bodhi_metadata.cr.ZCK_COMPRESSION == 99)
+        try:
+            bodhi_metadata.insert_in_repo(99, self.tempcompdir, 'garbage', 'zck', '/dev/null')
+        except TypeError:
+            pass
+        try:
+            bodhi_metadata.insert_in_repo(bodhi_metadata.cr.XZ_COMPRESSION, self.tempcompdir,
+                                          'garbage', 'zck', '/dev/null')
+        except TypeError:
+            pass
+        del bodhi_metadata.cr.ZCK_COMPRESSION
+        assert(not hasattr(bodhi_metadata.cr, 'ZCK_COMPRESSION'))
+        try:
+            bodhi_metadata.insert_in_repo(bodhi_metadata.cr.XZ_COMPRESSION, self.tempcompdir,
+                                          'garbage', 'xz', '/dev/null')
+        except TypeError:
+            pass
```

Comment:
We should add assertions that the right calls were made when the zchunk compression was available, otherwise all we know is that it didn't blow up.

Reply:
insert_in_repo() is called by UpdateInfoMetadata(), which is tested by _test_extended_metadata(), and I piggybacked the zchunk tests there.  UpdateInfoMetadata() creates both the regular and zchunked metadata, _verify_updateinfos() validates that we have two updateinfos if zchunk is enabled, and _test_extended_metadata() validates that both updateinfo files match what they're supposed to be.

So far, test_zchunk_metadata is a bit of a misnomer as all it does is make sure that we reach 100% code coverage, and doesn't really test anything.

Is there other testing we should be doing to verify that the zchunk compression is working?

Reply:
It's true that mock testing isn't nearly as valuable as testing that actually uses the code/resources you are mocking, but there is still *something* being tested, which is that the code is at least valid Python. Since Python is not a static language, mocking is a way to make sure that the code we have doesn't have problems like typos or things of that nature.